### DISCUSSION TITLE
recompile opencv for CUDA backend

### DISCUSSION THREAD

**Discussion by JustLuckyRobotics** [cyberbotics/wrestling#41] [Ideas]
Considering that the system that our code is ran on is given a Nvidia GPU, is it possible to get openCV recompiled to support CUDA backend. Otherwise it limits the possibilities for CV applications. 

We have already tested to see if the current library is built to support CUDA and the : "not built to support CUDA back end, switching to CPU" while this still works, we have found on local systems that it greatly hinders the simulation speed of our robot. 

i gather this might be a challenge depending on the setup and might not be a fast change to implement, however it could potentially benefit everyone.

**Comment by Jean-Eudes-le-retour** [comment]
The participants entries are run in [Docker containers](https://github.com/cyberbotics/wrestling#include-dependencies-in-your-docker-container). I have never recompiled openCV to support CUDA, but if you know the steps to configure a Linux machine to do so, it might be possible to write a Dockerfile that runs all the installation and compilation commands needed.

If you manage to write a working Dockerfile it would be great if you shared it somewhere!

**Reply by JustLuckyRobotics** [reply]
sure thing. this would require two things for the host machine however:

-correct Nvidia driver, for cuda 11.x the installed driver version needs to be >= 450.80.02 
-and installation of the Nvidia container toolkit and configuring docker to recognise it, instructions can be found here: https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html#docker

i believe these are things not possible from our end simply with the docker container stuff. otherwise this should be possible.

**Comment by omichel** [comment]
I believe we are missing some Docker options to allow GPU acceleration for robot controllers.
I just added them in https://github.com/cyberbotics/competition-record-action/pull/26.

**Reply by JustLuckyRobotics** [reply]
i just replied to jean, does this mean i no longer need configure the dockerfile?

**Reply by omichel** [reply]
No, you should still configure the Dockerfile to base it on a nvidia cuda image. But in addition to this, we have to configure the server, so that it can access the GPU and CUDA through Docker. We will work this out in https://github.com/cyberbotics/competition-record-action/pull/26 and let you know when it is ready for testing with a Dockerfile based on nvidia/cuda that you should define on your end.

**Reply by JustLuckyRobotics** [reply]
ok thank you.

1 concern i do have is with the compilation time, building opencv for cuda can often take upwards of 20-30 minutes. if it is rebuilt with every commit this could cause problems locking the comp down, especially when things start getting more boisterous and every team is committing multiple time an hour.

i believe i have what is needed to configure the dockerfile to test this when the docker is ready, can you confirm which version of linux is installed? i dont see it stated anywhere.

**Reply by omichel** [reply]
The host machine is Ubuntu 22.04. Normally, your docker image will be cached, including the build stage for opencv, so it should impact the server only the first time it is used. Alternatively, you may build your own docker image and host it on dockerhub.

**Reply by omichel** [reply]
This is merged. Now when I test it on the server machine, I am getting the correct output (see below).
I guess you should now be able to run a GPU enabled OpenCV controller. Let me know if you encounter any difficulty.
```bash
~$ docker run --gpus all --rm nvidia/cuda:11.6.2-base-ubuntu20.04 nvidia-smi
Mon Feb 27 10:16:13 2023
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.78.01    Driver Version: 525.78.01    CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  Off  | 00000000:01:00.0  On |                  N/A |
| 29%   30C    P8     6W / 120W |    483MiB /  3072MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
```
